### PR TITLE
Update errors/warnings/debug to user proper interfaces

### LINKED ISSUE #355: support more than 999 PEs in debug printout
It is possible that someone will run SOS with 1000+ PEs and want to see well-formed error message labels.  `%03d` limits the user to 999 PEs.

```
src/collectives.c:            fprintf(stderr, "[%03d] Bad barrier algorithm %s\n",
src/collectives.c:            fprintf(stderr, "[%03d] Bad broadcast algorithm %s\n",
src/collectives.c:            fprintf(stderr, "[%03d] Bad reduction algorithm %s\n",
src/collectives.c:            fprintf(stderr, "[%03d] Bad collect algorithm %s\n",
src/collectives.c:            fprintf(stderr, "[%03d] Bad fcollect algorithm %s\n",
src/init.c:                "[%03d] ERROR: SHMEM_VENDOR_STRING length (%zu) exceeds SHMEM_MAX_NAME_LEN (%d)\n",
src/init.c:                "[%03d] ERROR: symmetric heap initialization failed: %d\n",
src/init.c:                "[%03d] ERROR: Transport init failed\n",
src/init.c:                "[%03d] ERROR: XPMEM init failed\n",
src/init.c:                "[%03d] ERROR: CMA init failed\n",
src/init.c:        fprintf(stderr, "[%03d] ERROR: runtime exchange failed: %d\n", 
src/init.c:                "[%03d] ERROR: Transport startup failed\n",
src/init.c:                "[%03d] ERROR: XPMEM startup failed\n",
src/init.c:                "[%03d] ERROR: CMA startup failed\n",
src/init.c:                "[%03d] ERROR: initialization of collectives failed: %d\n",
src/malloc.c:                "[%03d] ERROR: symmetric heap usage error detected, possibly at 0x%lx.\n", \
src/malloc.c:                "[%03d] ERROR: symmetric heap data structure corruption found.\n", \
src/shmem_collectives.h:        fprintf(stderr, "[%03d] Illegal barrier type %d\n", 
src/shmem_collectives.h:        fprintf(stderr, "[%03d] Illegal broadcast type %d\n", 
src/shmem_collectives.h:        fprintf(stderr, "[%03d] Illegal reduction type %d\n", 
src/shmem_collectives.h:        fprintf(stderr, "[%03d] Illegal collect type %d\n", 
src/shmem_collectives.h:        fprintf(stderr, "[%03d] Illegal fcollect type %d\n", 
src/shmem_internal.h:        fprintf(stderr, "[%03d] WARN: %s:%d return code %d\n",         \
src/shmem_internal.h:        fprintf(stderr, "[%03d] ERROR: %s:%d return code %d\n",         \
src/shmem_internal.h:        fprintf(stderr, "[%03d] ERROR: %s:%d: %s\n",                    \
src/shmem_internal.h:        fprintf(stderr, "[%03d] WARN: %s:%d: %s\n",                    \
src/shmem_internal.h:        off = snprintf(str, sizeof(str), "[%03d] ERROR: %s:%d:\n",      \
src/shmem_internal.h:        off+= snprintf(str+off, sizeof(str)-off, "[%03d]        ",      \
src/shmem_internal.h:        off = snprintf(str, sizeof(str), "[%03d] WARN: %s:%d:\n",       \
src/shmem_internal.h:        off+= snprintf(str+off, sizeof(str)-off, "[%03d]       ",       \
src/shmem_internal.h:            fprintf(stderr, "[%03d] DEBUG: %s:%d: %s\n",                \
src/shmem_internal.h:            off = snprintf(str, sizeof(str), "[%03d] DEBUG: %s:%d:\n",  \
src/shmem_internal.h:            off+= snprintf(str+off, sizeof(str)-off, "[%03d]        ",  \
src/shmem_internal.h:        fprintf(stderr, "[%03d] Assertion Failed: %s:%d: %s\n",         \
src/symmetric_heap_c.c:        fprintf(stderr, "[%03d] WARNING: symmetric heap pointer pushed below start\n",
src/symmetric_heap_c.c:        fprintf(stderr, "[%03d] WARNING: Out of symmetric memory, heap size %ld, overrun %"PRIdPTR"\n"
src/symmetric_heap_c.c:                        "[%03d]          Try increasing SHMEM_SYMMETRIC_SIZE\n",
src/symmetric_heap_f.c:            fprintf(stderr, "[%03d] ERROR: shpalloc failure (invalid length).  Aborting job.\n",
src/symmetric_heap_f.c:            fprintf(stderr, "[%03d] ERROR: shpalloc failure.  Aborting job.\n",
src/symmetric_heap_f.c:            fprintf(stderr, "[%03d] shpclmove failure (invalid length)\n",
src/symmetric_heap_f.c:            fprintf(stderr, "[%03d] shpclmove failure\n",
src/transport_cma.c:        fprintf(stderr, "[%03d] %s() ERROR: runtime_put failed: %d\n",
src/transport_cma.c:        fprintf(stderr, "[%03d] ERROR: Too many local ranks for CMA transport.\n",
src/transport_cma.c:            fprintf(stderr, "[%03d] %s() ERROR: runtime_get failed: %d\n",
src/transport_cma.h:            printf("[%03d] ERROR: %s (0x%lx) outside of symmetric areas\n", \
src/transport_ofi.h:        RAISE_ERROR_MSG("[%03d] ERROR in %s: address (0x%p) outside of symmetric areas\n",
src/transport_ofi.h:        RAISE_ERROR_MSG("[%03d] ERROR in %s: address (0x%p) outside of symmetric areas\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlNIInit failed: %d\n",
src/transport_portals4.c:                "[%03d] ERROR: Remote virtual addressing feature enabled, but Portals\n"
src/transport_portals4.c:                    "[%03d] ERROR: Total data ordering feature enabled, but Portals\n"
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlGetPhysId failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: runtime_put failed: %d\n",
src/transport_portals4.c:            fprintf(stderr, "[%03d] ERROR: runtime_put failed: %d\n",
src/transport_portals4.c:            fprintf(stderr, "[%03d] ERROR: runtime_put failed: %d\n",
src/transport_portals4.c:            fprintf(stderr, "[%03d] ERROR: heap base address does not match with rank %03d and virtual addressing is enabled\n",
src/transport_portals4.c:            fprintf(stderr, "[%03d] ERROR: data base address does not match with rank %03d and virtual addressing is enabled\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlGetPhysId failed: %d\n",
src/transport_portals4.c:            fprintf(stderr, "[%03d] ERROR: runtime_get failed: %d\n",
src/transport_portals4.c:                fprintf(stderr, "[%03d] ERROR: Too many local ranks.\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlSetMap failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlGetUid failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: Max volatile size found to be %lu, too small to continue\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: Max atomic size found to be %lu, too small to continue\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: Max fetch atomic size found to be %lu, too small to continue\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlEQAlloc failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlPTAlloc of table entry failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlPTAlloc of data table failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlPTAlloc of heap table failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlCTAlloc of target ct failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlLEAppend of all memory failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlLEAppend of heap section failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlLEAppend of data section failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlCTAlloc of put ct failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlCTAlloc of get ct failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlMDBind of put MD failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlMDBind of put MD failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlMDBind of put cntr MD failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] ERROR: PtlMDBind of get MD failed: %d\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] WARNING: put count mismatch: %ld, %ld\n",
src/transport_portals4.c:        fprintf(stderr, "[%03d] WARNING: get count mismatch: %ld, %ld\n",
src/transport_portals4.h:            printf("[%03d] ERROR: target (0x%lx) outside of symmetric areas\n", \
src/transport_portals4.h:            printf("[%03d] ERROR: target (0x%lx) outside of symmetric areas\n", \
src/transport_portals4.h:        fprintf(stderr, "[%03d] ERROR: PtlPTAlloc of data table failed: %d\n",
src/transport_portals4.h:        fprintf(stderr, "[%03d] ERROR: data portal table index mis-match: "
src/transport_portals4.h:        fprintf(stderr, "[%03d] ERROR: PtlLEAppend of data section failed: %d\n",
src/transport_xpmem.c:        fprintf(stderr, "[%03d] ERROR: xpmem_make failed: %s\n",
src/transport_xpmem.c:        fprintf(stderr, "[%03d] ERROR: xpmem_make failed: %s\n",
src/transport_xpmem.c:        fprintf(stderr, "[%03d] ERROR: runtime_put failed: %d\n",
src/transport_xpmem.c:                fprintf(stderr, "[%03d] ERROR: runtime_get failed: %d\n",
src/transport_xpmem.c:                fprintf(stderr, "[%03d] ERROR: could not get data apid: %s\n",
src/transport_xpmem.c:                fprintf(stderr, "[%03d] ERROR: could not get data segment: %s\n",
src/transport_xpmem.c:                fprintf(stderr, "[%03d] ERROR: could not get heap apid: %s\n",
src/transport_xpmem.c:                fprintf(stderr, "[%03d] ERROR: could not get data segment: %s\n",
src/transport_xpmem.h:        fprintf(stderr, "[%03d] ERROR: target (0x%lx) outside of symmetric areas\n",
src/transport_xpmem.h:        fprintf(stderr, "[%03d] ERROR: target (0x%lx) outside of symmetric areas\n",
```

### PR DESCRIPTION
**Pull request by jdinan**
Lots of error message, warning, and debug messaging improvements.

Closes #355

### CODE REVIEW

**Comment by jdinan** [comment]
Ready for review.  Sorry, this one got a little longer than expected.

File: src/transport_ofi.c [hunk 1]
```diff
@@ -681,8 +681,8 @@ int atomicvalid_rtncheck(int ret, int atomic_size,
 {
     if ((ret != 0 || atomic_size == 0) && atomic_sup != ATOMIC_SOFT_SUPPORT) {
         if (atomic_sup == ATOMIC_WARNINGS) {
-            fprintf(stderr, "Warning OFI detected no support for atomic '%s' "
-                    "on type '%s'\n", strOP, strDT);
+            RAISE_WARN_MSG("Warning OFI detected no support for atomic '%s' "
+                           "on type '%s'\n", strOP, strDT);
```
Comment (wrrobin) [inline src/transport_ofi.c]: The word "Warning" in the message is redundant.

File: src/transport_portals4.c [hunk 2]
```diff
@@ -271,27 +271,28 @@ shmem_transport_init(long eager_size)
...
 #ifdef ENABLE_REMOTE_VIRTUAL_ADDRESSING
     if ((PTL_TARGET_BIND_INACCESSIBLE & ni_limits.features) == 0) {
-        fprintf(stderr,
-                "[%03d] ERROR: Remote virtual addressing feature enabled, but Portals\n"
-                "doesn't support PTL_TARGET_BIND_INACCESSIBLE.  Aborting.\n",
-                shmem_internal_my_pe);
+        RETURN_ERROR_MSG("Remote virtual addressing feature enabled, but Portals\n"
+                         RAISE_PE_PREFIX
+                         "doesn't support PTL_TARGET_BIND_INACCESSIBLE.  Aborting.\n",
+                         shmem_internal_my_pe);
```
Comment (wrrobin) [inline src/transport_portals4.c]: Sending shmem_internal_my_pe is not necessary.

File: src/transport_portals4.c [hunk 3]
```diff
@@ -271,27 +271,28 @@ shmem_transport_init(long eager_size)
...
     if ((PTL_TOTAL_DATA_ORDERING & ni_limits.features) == 0) {
         if (1 == WANT_TOTAL_DATA_ORDERING) {
-            fprintf(stderr,
-                    "[%03d] ERROR: Total data ordering feature enabled, but Portals\n"
-                    "doesn't support PTL_TOTAL_DATA_ORDERING.  Aborting.\n",
-                    shmem_internal_my_pe);
+            RETURN_ERROR_MSG("Total data ordering feature enabled, but Portals\n"
+                             RAISE_PE_PREFIX
+                             "doesn't support PTL_TOTAL_DATA_ORDERING.  Aborting.\n",
+                             shmem_internal_my_pe);
```
Comment (wrrobin) [inline src/transport_portals4.c]: Sending shmem_internal_my_pe is not necessary.

File: src/init.c [hunk 4]
```diff
@@ -106,7 +106,7 @@ shmem_internal_shutdown_atexit(void)
     if ( shmem_internal_initialized && !shmem_internal_finalized &&
          !shmem_internal_initialized_with_start_pes && !shmem_internal_global_exit_called &&
          shmem_internal_my_pe == 0) {
-        fprintf(stderr, "Warning: shutting down without a call to shmem_finalize()\n");
+        RAISE_WARN_STR("shutting down without a call to shmem_finalize");
```
Comment (jdinan) [inline src/init.c]: Calling ```shmem_finalize``` is not compulsory in OpenSHMEM 1.3, so this is the right way to handle it for now.  We'll need to revisit this when we implement 1.4; we may want to change this to ```RETURN_ERROR_STR```.

File: src/init.c [hunk 5]
```diff
@@ -213,27 +212,21 @@ shmem_internal_init(int tl_requested, int *tl_provided)
     /* create symmetric heap */
     ret = shmem_internal_symmetric_init(heap_size, heap_use_malloc);
     if (0 != ret) {
-        fprintf(stderr,
-                "[%03d] ERROR: symmetric heap initialization failed: %d\n",
-                shmem_internal_my_pe, ret);
+        RETURN_ERROR_MSG("symmetric heap initialization failed: %d\n", ret);
```
Comment (jdinan) [inline src/init.c]: There isn't really an explanation; it's just for debugging so we can trace back the cause of the error.

File: src/init.c [hunk 6]
```diff
@@ -253,44 +244,35 @@ shmem_internal_init(int tl_requested, int *tl_provided)
     /* exchange information */
     ret = shmem_runtime_exchange();
     if (0 != ret) {
-        fprintf(stderr, "[%03d] ERROR: runtime exchange failed: %d\n",
-                shmem_internal_my_pe, ret);
```
Comment (jdinan) [inline src/init.c]: Same as above.

File: src/init.c [hunk 7]
```diff
@@ -253,44 +244,35 @@ shmem_internal_init(int tl_requested, int *tl_provided)
...
         goto cleanup;
     }
 #endif
 
     ret = shmem_internal_collectives_init(crossover, radix);
     if (ret != 0) {
-        fprintf(stderr,
-                "[%03d] ERROR: initialization of collectives failed: %d\n",
-                shmem_internal_my_pe, ret);
+        RETURN_ERROR_MSG("initialization of collectives failed: %d\n", ret);
```
Comment (jdinan) [inline src/init.c]: Same as above; it's just for tracing back the source of the error.

File: src/shmem_collectives.h [hunk 8]
```diff
@@ -187,8 +187,8 @@ shmem_internal_op_to_all(void *target, const void *source, int count,
                                                pWrk, pSync, op, datatype);
             break;
     default:
-        fprintf(stderr, "[%03d] Illegal reduction type %d\n",
-                shmem_internal_my_pe, shmem_internal_reduce_type);
+            RAISE_ERROR_MSG("Illegal reduction type (%d)\n",
```
Comment (jdinan) [inline src/shmem_collectives.h]: Fixed

File: src/transport_ofi.c (hunk 1 above)
Comment (jdinan) [inline src/transport_ofi.c]: Fixed.  Actually, there are a bunch of these in OFI.  I'll fix them all.

File: src/transport_portals4.c (hunk 2 above)
Comment (jdinan) [inline src/transport_portals4.c]: The ```RAISE_PE_PREFIX``` macro is a string that prints the PE id, so it is required.  This is a bit messy, but it seemed like the most reasonable way to allow multi-line strings in the _MSG macros.

Comment (wrrobin) [inline src/transport_portals4.c]: Ok. sorry that I missed it.

**Comment by jdinan** [comment]
@wrrobin Please review the new changes.

b8cd180 should close #397

File: src/transport_ofi.c [hunk 9]
```diff
@@ -655,14 +655,14 @@ int populate_mr_tables(void)
                                     &shmem_transport_ofi_target_heap_addrs[i],
                                     sizeof(uint8_t*));
             if (err) {
-                RAISE_WARN_STR("Error getting heap addr from runtime KVS");
+                RAISE_WARN_STR("Get of heap addr from runtime KVS failed");
```
Comment (wrrobin) [inline src/transport_ofi.c]: Should we change "addr" -> "address" just to unify with the other warning messages?

File: src/transport_ofi.c [hunk 10]
```diff
@@ -655,14 +655,14 @@ int populate_mr_tables(void)
...
                 return 1;
             }
             err = shmem_runtime_get(i, "fi_data_addr",
                                     &shmem_transport_ofi_target_data_addrs[i],
                                     sizeof(uint8_t*));
             if (err) {
-                RAISE_WARN_STR("Error getting data segment addr from runtime KVS");
+                RAISE_WARN_STR("Get of data segment addr from runtime KVS failed");
```
Comment (wrrobin) [inline src/transport_ofi.c]: Should we change "addr" -> "address" just to unify with the other warning messages?

File: src/transport_ofi.c [hunk 11]
```diff
@@ -898,7 +893,7 @@ int populate_av(void)
         if (strncmp(myephostname, ephostname, EPHOSTNAMELEN) == 0) {
             SHMEM_SET_RANK_SAME_NODE(i, num_on_node++);
             if (num_on_node > 255) {
-                RAISE_WARN_STR("ERROR: Too many local ranks");
+                RAISE_WARN_STR("Number of local ranks exceeds limit (255)");
```
Comment (wrrobin) [inline src/transport_ofi.c]: In other error messages, we have put the return value in the parenthesis. Should we put 255 out of parenthesis here? May be like this:  "Number of local ranks exceeds limit value of 255"

**Comment by jdinan** [comment]
Thanks, @davidozog!  @wrrobin I pushed another commit with the fixes you suggested.

**Comment by jdinan** [comment]
Thanks for help with reviewing and improving this PR!  This ended up being a pretty nice update to our error messages.  :neckbeard:

**Comment by wrrobin** [comment]
Agree. Happy to be a part of this.

**Pull request merged**